gnomad_constraint/utils/constraint.py

```diff
@@ -82,3 +89,95 @@ def prepare_ht_for_constraint_calculations(ht: hl.Table) -> hl.Table:
+
+    Variants where there are no coverage, a low-quality variant is observed, or a variant with allele frequency
+    above `max_af` cutoff is observed in `exome_ht` and `context_ht` will be removed. Variants that are not synonymous or canonical will be
+    removed as well. For each substitution, context, methylation level, and exome coverage, the rest of variants in the `exome_ht` was counted
+    and annotated as `observed_variants`, and the rest of variants in the `context_ht` was counted and annotated as `possible_variants`. The
+    final Table is the outer-join of the filtered `exome_ht` and `context_ht` with the `observed_variants` and `possible_variants`.
+
+    The returned Table includes the following annotations:
+        - context - trinucleotide genomic context
+        - ref - the reference allele
+        - alt - the alternate base
+        - methylation_level - methylation_level
+        - observed_variants - observed variant counts in `exome_ht`
+        - possible_variants - possible variant counts in `context_ht`
+        - downsampling_counts_{pop} - variant counts in downsamplings for populations in `pops`
+        - mu_snp - SNP mutation rate
+        - annotations added by `annotate_mutation_type`
+
+    :param exome_ht: Preprocessed exome Table.
+    :param context_ht: Preprocessed context Table.
+    :param mutation_ht: Preprocessed mutation rate Table.
+    :param max_af: Maximum allele frequency for a variant to be included in returned counts. Default is 0.001.
+    :param keep_annotations: Annotations to keep in the context Table.
+    :param pops: List of populations to use for downsampling counts. Default is ().
+    :param grouping: Annotations other than 'context', 'ref', 'alt', and `methylation_level` to group by when counting variants. Default is ('exome_coverage',).
+    :param partition_hint: Target number of partitions for aggregation. Default is 100.
+    :return: Table with observed variant and possible variant count.
+    """
+    # Allele frequency information for high-quality genotypes (GQ >= 20; DP >= 10; and AB >= 0.2 for het calls) in all release samples in gnomAD
+    freq_expr = exome_ht.freq[0]
+
+    # Set up the criteria that variants where there are no coverage, a low-quality variant is observed, or a variant with allele frequency
+    # above `max_af` cutoff is observed in `exome_ht` and `context_ht` will be removed
+    keep_criteria = (
+        (freq_expr.AC > 0) & exome_ht.pass_filters & (freq_expr.AF <= max_af)
+    )
+    keep_annotations += grouping
+
+    # Keep variants that are synonymous, canonical, and satisfied criteria above
+    # Count the observed variants in the entire Table and in each downsampling grouped by `keep_annotations`
+    ht = count_variants_by_group(
+        filter_vep_transcript_csqs(exome_ht.filter(keep_criteria)).select(
+            *list(keep_annotations) + ["freq", "pass_filters"]
```

Comment:
I add `pass_filters ` since the old code has it. Just double check it makes sense.